pkg/xds/experimental/client/client_test.go

```diff
@@ -0,0 +1,1126 @@
+					}
+
+					var gotResourceNames []string
+					var gotTypeUrl string
+					if sotw {
+						got := <-sendCh
+						gotResourceNames = got.GetResourceNames()
+						gotTypeUrl = got.GetTypeUrl()
+					} else {
+						got := <-sendDeltaCh
+						gotResourceNames = got.GetResourceNamesSubscribe()
+						gotTypeUrl = got.GetTypeUrl()
+					}
+
+					var wantResourceNames []string
+					if sotw {
+						wantResourceNames = append(tc.resourceNames, tc.currResourceNames...)
+						sort.Strings(wantResourceNames)
+						wantResourceNames = slices.Compact(wantResourceNames)
+					} else {
+						wantResourceNames = tc.resourceNames
+					}
+					sort.Strings(gotResourceNames)
+
+					if diff := cmp.Diff(wantResourceNames, gotResourceNames); diff != "" {
+						t.Errorf("mismatched resource names: +got, -want: %s", diff)
+					}
+					if typeUrl != gotTypeUrl {
+						t.Errorf("typeUrl = %s, want = %s", gotTypeUrl, typeUrl)
+					}
+				})
+			}
+		})
+	}
+}
+
+type fakeClientConn struct {
+	OnInvoke    func(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error
+	OnNewStream func(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error)
+}
+
+func (f *fakeClientConn) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error {
+	return f.OnInvoke(ctx, method, args, reply, opts...)
+}
+
+func (f *fakeClientConn) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	return f.OnNewStream(ctx, desc, method, opts...)
+}
+
+var _ grpc.ClientConnInterface = (*fakeClientConn)(nil)
+
+func TestClientConnectionRetry(t *testing.T) {
+	for _, retry := range []bool{true, false} {
+		t.Run(fmt.Sprintf("retry=%v", retry), func(t *testing.T) {
+			defer goleak.VerifyNone(t)
+
+			opts := *Defaults
+			opts.MinBackoff = time.Hour
+			opts.MaxBackoff = time.Hour
+			opts.RetryConnection = retry
+			c := NewClientWithOptions(slog.Default(), testNode(), &opts)
+			ctx, cancel := context.WithCancel(context.TODO())
+			defer cancel()
+
+			wantErr := fmt.Errorf("test err")
+			newStreamCh := make(chan bool)
+			defer close(newStreamCh)
+			fakeConn := &fakeClientConn{
+				OnNewStream: func(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+					newStreamCh <- true
+					return nil, wantErr
+				},
+			}
+
+			done := make(chan bool)
+			var gotError error
+			go func() {
+				defer close(done)
+				gotError = c.Run(ctx, fakeConn)
+			}()
+			<-newStreamCh
+			if retry {
+				cancel()
+				<-done
+			} else {
+				<-done
+				cancel()
+			}
+
+			if retry {
+				wantErr = context.Canceled
+			}
+			if !errors.Is(gotError, wantErr) {
+				t.Errorf("error = %v, want = %v", gotError, wantErr)
+			}
+		})
+	}
+}
+
+func TestDeltaAckAndNack(t *testing.T) {
```

Comment:
No testing of SOTW ACK and NACK?

Reply:
It looks to me SOTW NACK is not supported?